File: typescript/test-compat.mjs
/**
 * Test: dhi schemas are compatible with Zod workflows
 *
 * Users who use Zod in their app should be able to:
 * 1. Import our schemas and use z.infer<> for types
 * 2. Use .parse() and .safeParse() on API responses
 * 3. Compose our schemas with their own Zod schemas (z.extend, z.merge, z.pick, z.omit)
 * 4. Use .shape to access individual fields
 * 5. Use .partial() and .required()
 */

import { z } from "dhi";

// ── Re-define SDK schemas inline (same as src/index.ts) ──

const InsertResultSchema = z.object({
  success: z.boolean(),
  id: z.number().int(),
  namespace: z.string(),
  upserted: z.boolean(),
});

const SearchResultSchema = z.object({
  id: z.number().int(),
  score: z.number(),
  metadata: z.record(z.string(), z.any()).optional(),
});

const SearchResponseSchema = z.object({
  results: z.array(SearchResultSchema),
  count: z.number().int(),
  namespace: z.string(),
});

const VectorEntrySchema = z.object({
  id: z.number().int().positive(),
  vector: z.array(z.number()),
  metadata: z.record(z.string(), z.any()).optional(),
});

let passed = 0;
let failed = 0;

function test(name, fn) {
  try {
    fn();
    passed++;
    console.log(`  PASS  ${name}`);
  } catch (e) {
    failed++;
    console.log(`  FAIL  ${name}: ${e.message}`);
  }
}

function assert(cond, msg) {
  if (!cond) throw new Error(msg || "assertion failed");
}

console.log("\n=== dhi <> Zod Compatibility Tests (TypeScript SDK) ===\n");

// ── 1. .parse() works on valid data ──
test("parse() accepts valid InsertResult", () => {
  const result = InsertResultSchema.parse({
    success: true,
    id: 42,
    namespace: "default",
    upserted: false,
  });
  assert(result.success === true);
  assert(result.id === 42);
  assert(result.namespace === "default");
});

// ── 2. .parse() rejects invalid data ──
test("parse() rejects invalid InsertResult", () => {
  let threw = false;
  try {
    InsertResultSchema.parse({ success: "yes", id: "abc" });
  } catch (e) {
    threw = true;
    assert(e.constructor.name === "ZodError", `Expected ZodError, got ${e.constructor.name}`);
  }
  assert(threw, "Should have thrown");
});

// ── 3. .safeParse() returns success/error ──
test("safeParse() returns success on valid data", () => {
  const result = InsertResultSchema.safeParse({
    success: true,
    id: 1,
    namespace: "prod",
    upserted: true,
  });
  assert(result.success === true);
  assert(result.data.id === 1);
});

test("safeParse() returns error on invalid data", () => {
  const result = InsertResultSchema.safeParse({ success: 123 });
  assert(result.success === false);
  assert(result.error !== undefined);
});

// ── 4. Nested schema parsing (SearchResponse with nested SearchResult[]) ──
test("parse() handles nested schemas", () => {
  const resp = SearchResponseSchema.parse({
    results: [
      { id: 1, score: 0.95, metadata: { title: "hello" } },
      { id: 2, score: 0.87 },
    ],
    count: 2,
    namespace: "default",
  });
  assert(resp.results.length === 2);
  assert(resp.results[0].score === 0.95);
  assert(resp.results[0].metadata.title === "hello");
  assert(resp.results[1].metadata === undefined);
});

// ── 5. .shape access ──
test(".shape gives access to individual field schemas", () => {
  const idSchema = InsertResultSchema.shape.id;
  assert(idSchema !== undefined, ".shape.id should exist");
  const parsed = idSchema.parse(42);
  assert(parsed === 42);
});

// ── 6. Schema composition: .extend() ──
test(".extend() adds fields to schema", () => {
  const ExtendedResult = InsertResultSchema.extend({
    latencyMs: z.number(),
  });
  const result = ExtendedResult.parse({
    success: true,
    id: 1,
    namespace: "default",
    upserted: false,
    latencyMs: 12.5,
  });
  assert(result.latencyMs === 12.5);
});

// ── 7. Schema composition: .pick() ──
test(".pick() selects specific fields", () => {
  const IdOnly = InsertResultSchema.pick({ id: true, namespace: true });
  const result = IdOnly.parse({ id: 99, namespace: "test" });
  assert(result.id === 99);
  assert(result.namespace === "test");
  // Extra fields should be stripped
  assert(result.success === undefined);
});

// ── 8. Schema composition: .omit() ──
test(".omit() removes specific fields", () => {
  const NoUpsert = InsertResultSchema.omit({ upserted: true });
  const result = NoUpsert.parse({ success: true, id: 1, namespace: "default" });
  assert(result.success === true);
  assert(result.upserted === undefined);
});

// ── 9. .partial() makes all fields optional ──
test(".partial() makes fields optional", () => {
  const PartialResult = InsertResultSchema.partial();
  const result = PartialResult.parse({ id: 5 });
  assert(result.id === 5);
  assert(result.success === undefined);
});

// ── 10. .merge() combines two schemas ──
test(".merge() combines schemas", () => {
  const Extra = z.object({ source: z.string() });
  const Merged = InsertResultSchema.merge(Extra);
  const result = Merged.parse({
    success: true,
    id: 1,
    namespace: "default",
    upserted: false,
    source: "api",
  });
  assert(result.source === "api");
  assert(result.success === true);
});

// ── 11. Positive int constraint on VectorEntry ──
test("VectorEntry rejects non-positive id", () => {
  let threw = false;
  try {
    VectorEntrySchema.parse({ id: -1, vector: [0.1] });
  } catch {
    threw = true;
  }
  assert(threw, "Should reject negative id");
});

test("VectorEntry accepts valid entry", () => {
  const entry = VectorEntrySchema.parse({
    id: 1,
    vector: [0.1, 0.2, 0.3],
    metadata: { title: "test" },
  });
  assert(entry.id === 1);
  assert(entry.vector.length === 3);
});

// ── 12. Array of SearchResults ──
test("z.array(SearchResultSchema) parses array", () => {
  const arr = z.array(SearchResultSchema).parse([
    { id: 1, score: 0.9 },
    { id: 2, score: 0.8, metadata: { a: 1 } },
  ]);
  assert(arr.length === 2);
});

// ── Summary ──
console.log(`\n  Results: ${passed} passed, ${failed} failed\n`);
if (failed > 0) process.exit(1);

File: python/test_compat.py
"""
Test: dhi models are compatible with Pydantic workflows

Users who use Pydantic in their app should be able to:
1. Use model_validate() / model_dump() / model_dump_json()
2. Use model_json_schema() for JSON Schema generation
3. Inherit/extend our models
4. Use Field constraints
5. Use .model_fields to introspect
6. Mix with their own Pydantic models
"""

import json
import sys
from typing import Any, Dict, List, Optional

# Simulate what a user would do: import from emergentdb
sys.path.insert(0, ".")
from emergentdb import (
    InsertResult,
    BatchInsertResult,
    SearchResult,
    SearchResponse,
    DeleteResult,
)


passed = 0
failed = 0


def test(name, fn):
    global passed, failed
    try:
        fn()
        passed += 1
        print(f"  PASS  {name}")
    except Exception as e:
        failed += 1
        print(f"  FAIL  {name}: {e}")


print("\n=== dhi <> Pydantic Compatibility Tests (Python SDK) ===\n")


# ── 1. model_validate() on valid data ──
def test_model_validate():
    result = InsertResult.model_validate(
        {"success": True, "id": 42, "namespace": "default", "upserted": False}
    )
    assert result.success is True
    assert result.id == 42
    assert result.namespace == "default"


test("model_validate() accepts valid InsertResult", test_model_validate)


# ── 2. model_validate() rejects bad data ──
def test_model_validate_reject():
    threw = False
    try:
        InsertResult.model_validate({"success": "yes", "id": "abc"})
    except Exception:
        threw = True
    assert threw, "Should have thrown on invalid data"


test("model_validate() rejects invalid data", test_model_validate_reject)


# ── 3. model_dump() returns dict ──
def test_model_dump():
    result = InsertResult(success=True, id=1, namespace="prod", upserted=True)
    d = result.model_dump()
    assert isinstance(d, dict)
    assert d["success"] is True
    assert d["id"] == 1
    assert d["namespace"] == "prod"
    assert d["upserted"] is True


test("model_dump() returns dict", test_model_dump)


# ── 4. model_dump_json() returns JSON string ──
def test_model_dump_json():
    result = InsertResult(success=True, id=1, namespace="default", upserted=False)
    j = result.model_dump_json()
    assert isinstance(j, str)
    data = json.loads(j)
    assert data["success"] is True
    assert data["id"] == 1


test("model_dump_json() returns valid JSON", test_model_dump_json)


# ── 5. Default values work ──
def test_defaults():
    # namespace and upserted have defaults
    result = InsertResult.model_validate({"success": True, "id": 5})
    assert result.namespace == "default"
    assert result.upserted is False


test("Default field values work", test_defaults)


# ── 6. Nested model validation (SearchResponse with List[SearchResult]) ──
def test_nested():
    resp = SearchResponse.model_validate(
        {
            "results": [
                {"id": 1, "score": 0.95, "metadata": {"title": "hello"}},
                {"id": 2, "score": 0.87},
            ],
            "count": 2,
            "namespace": "default",
        }
    )
    assert len(resp.results) == 2
    assert resp.results[0].score == 0.95
    assert resp.results[0].metadata == {"title": "hello"}
    assert resp.results[1].metadata is None


test("Nested model validation (SearchResponse)", test_nested)


# ── 7. model_json_schema() generates JSON Schema ──
def test_json_schema():
    schema = InsertResult.model_json_schema()
    assert isinstance(schema, dict)
    assert "properties" in schema
    assert "success" in schema["properties"]
    assert "id" in schema["properties"]


test("model_json_schema() generates valid schema", test_json_schema)


# ── 8. Model inheritance/extension ──
def test_inheritance():
    class ExtendedResult(InsertResult):
        latency_ms: float = 0.0

    result = ExtendedResult.model_validate(
        {
            "success": True,
            "id": 1,
            "namespace": "default",
            "upserted": False,
            "latency_ms": 12.5,
        }
    )
    assert result.latency_ms == 12.5
    assert result.success is True


test("Model inheritance/extension works", test_inheritance)


# ── 9. model_fields introspection ──
def test_model_fields():
    fields = InsertResult.model_fields
    assert "success" in fields
    assert "id" in fields
    assert "namespace" in fields
    assert "upserted" in fields


test("model_fields introspection", test_model_fields)


# ── 10. Attribute access on model instances ──
def test_attribute_access():
    result = SearchResult(id=1, score=0.95, metadata={"tag": "a"})
    assert result.id == 1
    assert result.score == 0.95
    assert result.metadata == {"tag": "a"}


test("Attribute access on model instances", test_attribute_access)


# ── 11. BatchInsertResult with all fields ──
def test_batch_insert():
    result = BatchInsertResult.model_validate(
        {
            "success": True,
            "ids": [1, 2, 3],
            "count": 3,
            "namespace": "test-ns",
            "new_count": 2,
            "upserted_count": 1,
        }
    )
    assert result.ids == [1, 2, 3]
    assert result.count == 3
    assert result.namespace == "test-ns"
    assert result.new_count == 2
    assert result.upserted_count == 1


test("BatchInsertResult with all fields", test_batch_insert)


# ── 12. DeleteResult model ──
def test_delete():
    result = DeleteResult.model_validate(
        {"deleted": True, "id": 99, "namespace": "cleanup"}
    )
    assert result.deleted is True
    assert result.id == 99
    assert result.namespace == "cleanup"


test("DeleteResult model", test_delete)


# ── 13. model_validate from JSON string ──
def test_validate_json():
    json_str = '{"success": true, "id": 1, "namespace": "default", "upserted": false}'
    result = InsertResult.model_validate_json(json_str)
    assert result.success is True
    assert result.id == 1


test("model_validate_json() from JSON string", test_validate_json)


# ── 14. Can use models as type hints in user code ──
def test_type_hint():
    def process_results(response: SearchResponse) -> List[int]:
        return [r.id for r in response.results]

    resp = SearchResponse(
        results=[SearchResult(id=1, score=0.9), SearchResult(id=2, score=0.8)],
        count=2,
        namespace="default",
    )
    ids = process_results(resp)
    assert ids == [1, 2]


test("Models work as type hints", test_type_hint)


# ── Summary ──
print(f"\n  Results: {passed} passed, {failed} failed\n")
if failed > 0:
    sys.exit(1)
